Modix.Services/Moderation/MessageContentCheckBehaviour.cs

```diff
@@ -0,0 +1,134 @@
+            IModerationService moderationService, IMessageContentPatternService messageContentPatternService)
+        {
+            _designatedChannelService = designatedChannelService;
+            _discordSocketClient = discordSocketClient;
+            _authorizationService = authorizationService;
+            _moderationService = moderationService;
+            _messageContentPatternService = messageContentPatternService;
+        }
+
+        public Task HandleNotificationAsync(MessageReceivedNotification notification,
+            CancellationToken cancellationToken = default)
+            => TryCheckMessageAsync(notification.Message);
+
+        public Task HandleNotificationAsync(MessageUpdatedNotification notification,
+            CancellationToken cancellationToken = default)
+            => TryCheckMessageAsync(notification.NewMessage);
+
+        private async Task TryCheckMessageAsync(IMessage message)
+        {
+            if (!(message.Author is IGuildUser author)
+                || (author.Guild is null)
+                || !(message.Channel is IGuildChannel channel)
+                || (channel.Guild is null))
+            {
+                Log.Debug(
+                    "Message {MessageId} was not in an IGuildChannel & IMessageChannel, or Author {Author} was not an IGuildUser",
+                    message.Id, message.Author.Id);
+                return;
+            }
+
+            if (author.Id == _discordSocketClient.CurrentUser.Id)
+                return;
+
+            var isContentBlocked = await IsContentBlocked(channel, message);
+
+            if (!isContentBlocked)
+            {
+                return;
+            }
+
+            if (await _designatedChannelService.ChannelHasDesignationAsync(channel.Guild,
+                channel, DesignatedChannelType.Unmoderated, default))
+            {
+                return;
+            }
+
+            if (await _authorizationService.HasClaimsAsync(author, AuthorizationClaim.BypassMessageContentPatternCheck))
+            {
+                Log.Debug("Message {MessageId} was skipped because the author {Author} has the {Claim} claim",
+                    message.Id, message.Author.Id, AuthorizationClaim.BypassMessageContentPatternCheck);
+                return;
+            }
+
+            Log.Debug("Message {MessageId} is going to be deleted", message.Id);
+
+            await _moderationService.DeleteMessageAsync(message, "Unauthorized Message Content Link",
+                _discordSocketClient.CurrentUser.Id, default);
+
+            Log.Debug("Message {MessageId} was deleted because it contains blocked content", message.Id);
+
+            await message.Channel.SendMessageAsync($"Sorry {author.Mention} your link has been removed!");
+        }
+
+        private async Task<bool> IsContentBlocked(IGuildChannel channel, IMessage message)
+        {
+            var patterns = await _messageContentPatternService.GetPatterns(channel.GuildId);
```

Comment:
I'm hoping this won't be too bad, doing an extra database call every time someone posts a message. A good future optimization could be to cache the patterns in memory so we don't have to pull them constantly.